### ISSUE TITLE
Precisely specify the behaviour of timeout and implement it correctly in all targets.

### ISSUE THREAD

**Issue by cmnrd** [lf-lang/lingua-franca#1496]
While trying to fix https://github.com/lf-lang/lingua-franca/issues/725 for C++, I realized that we don't have a clear definition of timeout and targets implement this differently. The website says:
> The [target property timeout](https://www.lf-lang.org/docs/handbook/target-declaration#timeout) specifies the last logical time at which reactions should be triggered. The last invocation of reactions will be at tag (timeout, 0).

My interpretation of this sentence is that the shutdown reaction should be invoked precisely at tag (timeout, 0). 

In C++, currently the last tag is (timeout, 1). In C, (timeout, 0) is only the last tag if keepalive is enabled or there is another event scheduled at (timeout, 0). Otherwise, shutdown will be invoked at the last scheduled event with a tag smaller than (timeout, 0).

How is this implemented in the other targets and what should be the correct behavior?

**Comment by edwardalee** [comment]
To me, the word "timeout" implies a bound, a maximum. I think that if keepalive == false, it is ok to end execution early and invoke shutdown reactions. It would seem strange to me to keep a program running while it can't do anything. 

As for the micostep, since the time is known in advance, I see no reason to postpone execution to microstep 1.

**Comment by cmnrd** [comment]
The microstep 1 in C++ is an artifact of how timeout is currently implemented in C++. #725 already says that this should be changed.

> To me, the word "timeout" implies a bound, a maximum. I think that if keepalive == false, it is ok to end execution early and invoke shutdown reactions. It would seem strange to me to keep a program running while it can't do anything.

I agree to this for programs that do not schedule events beyond the specified timeout (For instance if there are only startup reactions or timers without a period). However, for programs that would run indefinitely without a timeout (like any program that uses a periodic timer), I would expect the shutdown reactions to be invoked precisely at the timeout. However, in C, a program with a timer t(500ms, 1s) and timeout=2s terminates at tag (1500ms, 0), which seems odd to me.

**Comment by lhstrh** [comment]
@edwardalee, I agree with the intuition, but the wording in the documentation doesn't convey this meaning. How about we rephrase it as follows:
> The [target property timeout](https://www.lf-lang.org/docs/handbook/target-declaration#timeout) specifies an upper bound on the logical execution time of the program. When `timeout` is set, the last event to be handled will occur at tag `(timeout, 0)` or at one microstep after the tag of the last event that was pulled from the event queue before it runs empty, whichever tag is smaller. To prevent the event queue from running empty, the `keepalive` target property can be set to `true`.

While writing this, I do feel that it's somewhat strange that when we hit the timeout, regular events at microstep `0` occur simultaneously with shutdown events, whereas when the queue runs empty, shutdown reactions will happen at some arbitrary microstep _after_ the last regular event was handled. This asymmetry is because we can't conclude that the queue will be empty until after execution all reactions at the current time.

**Comment by edwardalee** [comment]
How about this variant:

> The [target property timeout](https://www.lf-lang.org/docs/handbook/target-declaration#timeout) specifies an upper bound on the logical execution time of the program. When timeout is set, no event with tag greater than `(timeout, 0)` will be handled. If the `keepalive` target property is true, then the program will run until tag `(timeout, 0)`, and any reactions to `shutdown` will be invoked at that tag. If `keepalive` is `false`, then the program will terminate either at `(timeout, 0)` or at some `(t, n+1)`, where `(t, n)` is a tag earlier than `(timeout, 0)` when the event queue becomes empty (no more events are scheduled for tags greater than `(t, n)`. Reactions to `shutdown` will be invoked at `(t, n+1)`.

We had extensive discussions about the choice of `(timeout, 0)` vs. `(timeout, 1)`. One reason is that we selected `(timeout, 0)` is that we can't prevent a reaction to `shutdown` from producing an output, in which case it could trigger other reactions that are not reactions to `shutdown`. Thus, we can't assert that only reactions to `shutdown` execute at that tag. Secondly, we thought that since the programmer is choosing both the `timeout` time and the other events, they would be more likely to expect them to be simultaneous if they are specified to occur at the same time, e.g. a periodic timer.

**Comment by lhstrh** [comment]
Yes, this is much better. I tried to pack to much into too few sentences. Making `n` explicit makes it clearer.

**Comment by edwardalee** [comment]
On this from @cmnrd:

>However, in C, a program with a timer t(500ms, 1s) and timeout=2s terminates at tag (1500ms, 0), which seems odd to me. It is also confusing that specifying keepalive in combination with timeout=2s on a program that runs indefinetely changes the behavior to terminate at tag (2s, 0).

I guess one argument is that what the C target approach is, in some sense, "more general," because there are more possible behaviors. We can achieve the behavior you propose and more.  Just set `keepalive = true`. 

Of course, "more general" is not automatically a good argument for doing something. I would be OK with changing it so that reactions to shutdown are invoked exactly at `(timeout, 0)`, as long as I don't have to implement the change. Note, however, that, with this change, specifying a `timeout` makes `keepalive` meaningless. It is automatically true, just as if there were physical action.

**Comment by cmnrd** [comment]
I think there is still a bit of confusion here. Let me try to lay out the different cases that we talk about:

Let's assume that for all cases we have `timeout=2s` specified: 
1. For a program that has no events scheduled beyond a tag `(t, n)` with `t  < 2s` and that has set `keepalive = false`, the consensus is that the shutdown reactions should be invoked at tag `(t, n+1)`.
2. For the same program from 1. with `keepalive = true`, the consensus is that the shutdown reactions should be invoked at tag `(2s, 0)`.
3. For a program that schedules events beyond the given timeout (e.g. any program with a periodic timer) and that has `keepalive = false`, I would expect the shutdown reactions to be invoked at tag `(timeout, 0)`. In C, however, the shutdown reactions are invoked at tag `(t, n+1)`, where `(t, n)` is the last tag on the event queue with `t < 2s`.
4. In the case of 3. with `keepalive = true` there is again consensus that the shutdown reactions should be invoked at `(t, n+1)`.

>  Note, however, that, with this change, specifying a timeout makes keepalive meaningless. It is automatically true, just as if there were physical action.

For a program with a periodic timer and no timeout (case 3.), `keepalive` is already redundant. I don't understand why `keepalive` should influence the behavior of `timeout` in this case.

**Comment by cmnrd** [comment]
I did a bit of research, and it looks like all of our targets implement this differently. The TypeScript behavior is similar to the current C++ one. The shutdown reactions are always (even in case 1) invoked at tag `(t, n)`, but I don't know value of `n` as I don't know how to access the current microstep in TS.

In Rust, the shutdown reactions are always invoked at tag `(timeout, 0)`. However, in case 1. it appears that although the tag is `(timeout, 0)`, the shutdown reactions are executed earlier, immediately after the event queue becomes empty. Case 2. and 4. I could not check, as apparently Rust does not implement keepalive at the moment.

**Comment by edwardalee** [comment]
This is an excellent summary, thanks!  On this:

> For a program with a periodic timer (case 3.) and no timeout, keepalive is already redundant.

To me, this is expected. A periodic timer already keeps things alive.  Making keepalive irrelevant when there *is* a timeout is a bit more surprising for me. One way to interpret the combination a periodic timer and a timeout is that timer events beyond the timeout *do not exist*. Under that interpretation, the C behavior is exactly what I would expect.

One extension we have talked about is to specify a timeout (or a finite number of events) for each timer.  E.g., timer(0, 1s, 60) would fire the timer 60 times.  Such an extension would interact with timeout and keepalive exactly as done by the C target in case 3.

But as I said, I'm ok with the change if that is the consensus. We don't really lose expressivity (for unfederated programs) because we can achieve the current C behavior using request_stop(), although this requires a reactor to have a global view of pending events. For federated programs, it becomes impossible to achieve that behavior, but I don't think it's a big deal.

**Comment by cmnrd** [comment]
> One way to interpret the combination a periodic timer and a timeout is that timer events beyond the timeout do not exist. Under that interpretation, the C behavior is exactly what I would expect.

Yer, I think it is really this interpretation that makes the difference. For me, the events beyond the timeout are still there, they are just never processed (btw this is also how it is implemented).

I am still unsure how to move forward with this. Considering the C++ implementation. I would also like to keep the effort low and I am a bit afraid of the overhead introduced by the implementation of timeout. In the C++ runtime currently the bottleneck is the scheduling of new events. Implementing timeout as sketched above would require to compare the tag of each scheduled event to the timeout. While this should be a relative fast operation, it definitely adds to the critical path.

How should we move forward with this? I am inclined to say that the specification should only say that, if timeout is set, no event with a tag greater than `(timeout, 0)` will be processed, but leave it unspecified when shutdown is invoked. Then it is up to the runtimes to decide when to invoke shutdown in case of a timeout. Moreover, all our runtimes then  adhere to this specification (given a small change in C++ to invoke shutdown at `(timeout, 0)` instead of `(timeout, 1)`.

**Comment by oowekyala** [comment]
> How should we move forward with this? I am inclined to say that the specification should only say that, if timeout is set, no event with a tag greater than `(timeout, 0)` will be processed, but leave it unspecified when shutdown is invoked. Then it is up to the runtimes to decide when to invoke shutdown in case of a timeout. Moreover, all our runtimes then adhere to this specification (given a small change in C++ to invoke shutdown at `(timeout, 0)` instead of `(timeout, 1)`.

I think this proposed semantics make sense. In the Rust target the keepalive option is not needed as the runtime knows with certainty when the event queue will remain empty forever. In this case, it invokes the shutdown reaction as soon as possible (\*). If the user really wants the shutdown reaction to be invoked at a specific time step, they can use a timer that provokes a request_stop:

```
timer stop(2s)
// shutdown will be invoked at tag (2s, 1)
reaction (stop) {= ctx.request_stop(Asap); =}
```
Although there would be no way to make the runtime stop at some `(t, 0)`...

(\*) as Christian pointed out the tag is wrong in the current implementation. It should be `(t, n+1)`, where `(t, n)` is the first tag after whose processing the event queue is empty forever (and `t < timeout` if a timeout is given). This would be consistent with cases 1 and 3 in the summary above.

**Comment by edwardalee** [comment]
> In the Rust target the keepalive option is not needed as the runtime knows with certainty when the event queue will remain empty forever.

Do I infer correctly from this that the Rust runtime does not support physical actions?

**Comment by edwardalee** [comment]
I feel like it would be unfortunate to not converge on a single semantics. Perhaps it would be worth measuring the performance impact?

**Comment by oowekyala** [comment]
> > In the Rust target the keepalive option is not needed as the runtime knows with certainty when the event queue will remain empty forever.
> Do I infer correctly from this that the Rust runtime does not support physical actions?

They are supported, but the only way to send an asynchronous event to the scheduler is through a channel instance. Channels are reference counted, so if the scheduler knows that if there are no live channels at a point where the event queue is empty, then no part of the program is able to send a new event. Hence the queue is empty forever, and there is no reason to keep the scheduler alive.

**Comment by cmnrd** [comment]
What should we do with the specification of timeout? Given that `keepalive` is not needed in Rust and also is not supported anymore by C++, it seems like a bad idea to couple the behavior of `timeout` with `keepalive`.

This also came up in a recent discussion with @lsk567. It looks like the current C behavior for timeout significantly complicates static scheduling.

**Comment by edwardalee** [comment]
I think the complication in static scheduling has nothing to do with when the timeout occurs.  It only has to do with the fact at the timeout, the schedule is different than the steady-state periodic schedule.

Also, note that `keepalive` is used in the embedded labs, albeit only as a stepping stone to doing more interesting things.  See https://www.lf-lang.org/embedded-lab/Interrupts.html

**Comment by cmnrd** [comment]
Yes, the functionality of `keepalive` is important, but it is implemented implicitly in C++ and Rust. As long as there are physical actions that can produce events, the program will not terminate. However, I don't think the behavior of `timeout` should depend on the presence of physical actions (that would be fragile) and it is also unnecessarily complicated to implement for the C++ and Rust targets.

I wonder how we can align the timeout behavior between targets. The description of timeout on the website is incomplete and also incorrect for most targets (including C). It would be great if we can specify a common behavior that works for all the targets. If we cannot reach a consensus, we should at least correctly describe the target specific behavior for all targets on the website

**Comment by cmnrd** [comment]
Yes, the proposal is to always shutdown at tag (timeout, 0). Only if there are no periodic timers and the runtime runs out of events, it can shutdown earlier.

>  As I've stated repeatedly in this thread, I'm OK with that, but don't have the bandwidth right now to implement it.

Sorry, in my memory this discussion was still unresolved, but it looks like there is actually agreement. Its fine that there is a lack in bandwidth. Most importantly I would like to have a clear specification. Then we can treat any misalignment in the targets as bugs.

>  I would be OK with changing it so that reactions to shutdown are invoked exactly at (timeout, 0), as long as I don't have to implement the change. Note, however, that, with this change, specifying a timeout makes keepalive meaningless. It is automatically true, just as if there were physical action.

This implies that a simple mechanism for implementing the new behavior, would be to automatically set `keepalive` whenever there is a `timeout` specified. Right?


>  I'm also OK with eliminating keepalive, although this will require updating the lab exercise (which is minor... I can do that once the change is implemented).

Note that this is not about eliminating `keepalive`, it is just about disentangling the influence of `keepalive` on `timeout`. To me, handling `keepalive` is a separate issue and it is also completely Ok for me to keep it in the C target.

**Comment by edwardalee** [comment]
> Yes, the proposal is to always shutdown at tag (timeout, 0). Only if there are no periodic timers and the runtime runs out of events, it can shutdown earlier.
> 
> > As I've stated repeatedly in this thread, I'm OK with that, but don't have the bandwidth right now to implement it.
> 
> Sorry, in my memory this discussion was still unresolved, but it looks like there is actually agreement. Its fine that there is a lack in bandwidth. Most importantly I would like to have a clear specification. Then we can treat any misalignment in the targets as bugs.
> 
> > I would be OK with changing it so that reactions to shutdown are invoked exactly at (timeout, 0), as long as I don't have to implement the change. Note, however, that, with this change, specifying a timeout makes keepalive meaningless. It is automatically true, just as if there were physical action.
> 
> This implies that a simple mechanism for implementing the new behavior, would be to automatically set `keepalive` whenever there is a `timeout` specified. Right?
> 
> > I'm also OK with eliminating keepalive, although this will require updating the lab exercise (which is minor... I can do that once the change is implemented).
> 
> Note that this is not about eliminating `keepalive`, it is just about disentangling the influence of `keepalive` on `timeout`. To me, handling `keepalive` is a separate issue and it is also completely Ok for me to keep it in the C target.

Ah, this is a good idea.  While we are at it, I would like to get rid of the warning when `keepalive` is set to true because of a physical action. This is not an interesting warning.
